data/txHandler.go

```diff
@@ -239,12 +252,39 @@ func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) net
 			logging.Base().Warnf("Received a non-decodable txn: %v", err)
 			return network.OutgoingMessage{Action: network.Disconnect}
 		}
+		decTxGroupElemOffsets[ntx] = dec.Consumed()
 		ntx++
+		if ntx >= config.MaxTxGroupSize {
+			// max group size reached, exit
+			// TODO: add metric after #4786 merged
+			break
+		}
 	}
 	if ntx == 0 {
 		logging.Base().Warnf("Received empty tx group")
 		return network.OutgoingMessage{Action: network.Disconnect}
 	}
+
+	// calculate hashes and check duplicates.
+	allTxSeen := true
+	start := 0
+	for i := 0; i < ntx; i++ {
+		end := decTxGroupElemOffsets[i]
+		rawtxn := rawmsg.Data[start:end]
+		start = end
+
+		d := crypto.Hash(rawtxn)
```

Comment:
This is a very expensive operation. We need to evaluate the impact of this before moving forward. 
We are adding more time on the single-threaded pipeline before dispersing the rest of the work to the exec pool, where the operations are in parallel.

Reply:
I agree, this is expensive but this is called per peer incoming peer thread.
The issue is we want 1) check dupes before the pool 2) to be fast 3) to be collision free to prevent DOS

Reply:
We could theoretically use a faster hash algorithm instead of SHA512_256 but would likely need to import a new library for that

Reply:
now we have two caches - fast blake2b with salt for raw messages, and slow SHA512_256 (that is faster than SHA256 on modern hardware) for canonical representation. The first helps filtering out normal duplicates, the second prevents txn duplicates if someone tries to trick the first one.
Either can be disabled with config options